HPCC-34238 Enhance IThreadPool for pool size to be changed on the fly

streeterd opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

streeterd committed:
HPCC-34238 Enhance IThreadPool for pool size to be changed on the fly

Implement new setPoolSize function for IThreadPool ensuring that
running threads are not affected.

Signed-off-by: Dave Streeter <dave.streeter@lexisnexisrisk.com>

streeterd committed:
HPCC-34258 Stopped threads are releasing unacquired slots

Prevent a thread allocated from a thread pool during dynamic bursting above
the configured maximum from persistently raising that maximum.

Signed-off-by: Dave Streeter <dave.streeter@lexisnexisrisk.com>

jakesmith reviewed (changes requested):
@streeterd - please see comments, but might be worth discussing further before before proceeding.

streeterd committed:
1. newDefaultmax approach removed.
2. slotsToRelease holds the number of slots to release after setPoolSize.
3. slotsToRelease is decremented for unused slots in setPoolSize.
4. slotsToRelease is decremented for any running threads that would not have
   released slots but now will.
5. slotsToRelease is decremented in _start for any unused slots.

Signed-off-by: Dave Streeter <dave.streeter@lexisnexisrisk.com>

streeterd commented on system/jlib/jthread.cpp:
```diff
@@ -832,6 +832,8 @@ protected: friend class CPooledThreadWrapper;
     std::atomic<bool> stopall{false};
     const bool inheritThreadContext;
     unsigned defaultmax = 0;
+    unsigned newDefaultmax = 0;
```
Fixed in b4b98aa7e36ff3f279242cbe5fb8aacd32622f7d

streeterd commented on system/jlib/jthread.cpp:
```diff
@@ -1299,8 +1312,31 @@ class CThreadPool: public CThreadPoolBase, implements IThreadPool, public CInter
 
+        if (newPoolSize < defaultmax)
+        {
+            if (!runningCount())
```
Fixed in b4b98aa7e36ff3f279242cbe5fb8aacd32622f7d

streeterd commented:
> @streeterd - please see comments, but might be worth discussing further before before proceeding.

This PR is now based on https://github.com/hpcc-systems/HPCC-Platform/pull/19941 to take advantage of the `slotConsumed` code.

streeterd committed:
Only check for running threads if runningCount() is non-zero

Signed-off-by: Dave Streeter <dave.streeter@lexisnexisrisk.com>

jakesmith commented on system/jlib/jthread.cpp:
```diff
@@ -1299,8 +1334,42 @@ class CThreadPool: public CThreadPoolBase, implements IThreadPool, public CInter
+        CriticalBlock block(crit);
+        if (newPoolSize > defaultmax)
+        {
+            availsem.signal(newPoolSize - defaultmax);
```
What if there are already 'burst' threads running?
Simply increasing the # of slots in availsem, will immediately allow new blocked threads to start
but if burst threads are already running, they should be considered as already occupying a slot.

The upshot of not taking into account already running threads when increasing the limit, is that increaing the pool size can immediately mean more new threads launch.

To test could:
1) setting limit to e.g. 2
2) starting 4 (with small pool delay), i.e. 2 are 'bursting' threads
3) setPoolSize(4)
4) start 2 more pool threads
5) After v. short delay, check how many running. We want there to be 4 thread running, but there will be 6.

jakesmith commented on system/jlib/jthread.cpp:
```diff
@@ -1299,8 +1334,42 @@ class CThreadPool: public CThreadPoolBase, implements IThreadPool, public CInter
+                    break; // none available
+                --slotsToRelease;
+            }
+            if (slotsToRelease && runningCount())
```
If while loop above failed to decrease by amount needed, then I think can only be true if some are running, so should assert.

jakesmith commented on system/jlib/jthread.cpp:
```diff
@@ -1299,8 +1334,42 @@ class CThreadPool: public CThreadPoolBase, implements IThreadPool, public CInter
+                        if (!--slotsToRelease)
+                            break;
+                    }
+                }
```
how can slotsToConsume be >0 after this for loop?

jakesmith commented on system/jlib/jthread.hpp:
```diff
@@ -364,6 +365,7 @@ class jlib_decl PerfTracer
 private:
     void dostart(unsigned seconds);
     void dostop();
+    virtual void setPoolSize(unsigned newPoolSize) { UNIMPLEMENTED;}
```
why is this here?

jakesmith commented on system/jlib/jthread.cpp:
```diff
@@ -1299,8 +1334,42 @@ class CThreadPool: public CThreadPoolBase, implements IThreadPool, public CInter
         return false;
     }
-};
+    void setPoolSize(unsigned newPoolSize)
```
missing 'virtual' and 'override'

jakesmith commented on system/jlib/jthread.cpp:
```diff
@@ -1003,16 +1020,29 @@ class CThreadPool: public CThreadPoolBase, implements IThreadPool, public CInter
+        bool slotConsumed{true};
         if (defaultmax)
         {
+            if (slotsToRelease)
```
let's discuss, but I don't think code in _start is needed/correct.

jakesmith reviewed (commented):
@streeterd - please see comments.

streeterd committed:
HPCC-34238 PR WIP

Signed-off-by: Dave Streeter <dave.streeter@lexisnexisrisk.com>
